pkg/controller/toolchainstatus/toolchainstatus_controller.go

```diff
@@ -0,0 +1,438 @@
+func (r *ReconcileToolchainStatus) membersHandleStatus(reqLogger logr.Logger, toolchainStatus *toolchainv1alpha1.ToolchainStatus) (componentError error) {
+	// get member clusters
+	memberClusters := r.getMembersFunc()
+	members := []toolchainv1alpha1.Member{}
+	if len(memberClusters) == 0 {
+		err := fmt.Errorf("no member clusters found")
+		memberStatusNotFoundCondition := status.NewComponentErrorCondition(toolchainv1alpha1.ToolchainStatusMemberStatusNoClustersFoundReason, err.Error())
+		memberStatus := customMemberStatus(*memberStatusNotFoundCondition)
+		noMemberClusterStatus := toolchainv1alpha1.Member{
+			ClusterName:  "",
+			MemberStatus: memberStatus,
+		}
+		members = append(members, noMemberClusterStatus)
+		return err
+	}
+	for _, cluster := range memberClusters {
+		memberClient := cluster.Client
+		if memberClient == nil {
+			// there is no client for the member cluster, create a status condition and add it to this member's status
+			err := fmt.Errorf("member cluster does not have a client assigned")
+			componentError = err
+			notFoundCondition := status.NewComponentErrorCondition(toolchainv1alpha1.ToolchainStatusMemberStatusNotFoundReason, err.Error())
+			memberStatus := customMemberStatus(*notFoundCondition)
+			members = append(members, memberResult(cluster, memberStatus))
+			continue
+		}
+		memberStatusObj := &toolchainv1alpha1.MemberStatus{}
+		err := memberClient.Get(context.TODO(), types.NamespacedName{Namespace: cluster.OperatorNamespace, Name: memberStatusName}, memberStatusObj)
+		if err != nil {
+			// couldn't find the memberstatus resource on the member cluster, create a status condition and add it to this member's status
+			wrappedErr := errs.Wrap(err, fmt.Sprintf("cannot find memberstatus resource in namespace %s", cluster.OperatorNamespace))
+			componentError = wrappedErr
+			memberStatusNotFoundCondition := status.NewComponentErrorCondition(toolchainv1alpha1.ToolchainStatusMemberStatusNotFoundReason, wrappedErr.Error())
+			memberStatus := customMemberStatus(*memberStatusNotFoundCondition)
+			members = append(members, memberResult(cluster, memberStatus))
+			continue
+		}
+
+		// take only the overall conditions from the memberstatus (ie. don't include status of member operator, host connection, etc.) in order to keep readability.
+		// member status details can be viewed on the member cluster directly.
+		memberStatus := toolchainv1alpha1.MemberStatusStatus{Conditions: memberStatusObj.Status.Conditions}
+		for _, condition := range memberStatusObj.Status.Conditions {
+			if condition.Type == toolchainv1alpha1.ConditionReady && condition.Status != "True" {
+				// the existing memberstatus condition is added to the toolchainstatus but the condition is not ready so set the component error to bubble up the error to the overall toolchain status
+				componentError = errs.Wrap(fmt.Errorf(condition.Message), fmt.Sprintf("member cluster %s not ready", cluster.Name))
```

Comment:
here `componentError` will only hold an error message for the last cluster which is not ready. Should we have a custom `ComponentError` type which may hold multiple cluster names, then return an error message with all the names at once for example?

Reply:
I see your point but componentError is mainly used to indicate that there is a problem with a specific component, in this case the Members portion of the toolchainstatus. The details of the error will still be visible in the status condition for each member. I suppose we don't even need to really return an error, it could be a bool but we currently log the error using the reqLogger here: https://github.com/rajivnathan/host-operator/blob/a388dad6a7aa9ca73a51f3da87a1721bfa012b58/pkg/controller/toolchainstatus/toolchainstatus_controller.go#L172

But your point still stands in that only the error for the last cluster would be logged. Is there a good way to combine multiple errors into a single error and still have it clear and readable? Otherwise, maybe it's okay to only log the last error since each error is visible in the toolchainstatus?

Reply:
I've seen functions that return an `Errors` instead of a standard `error`, for example.

Reply:
I think this is worth a bit more investigation and discussion so I'd like to handle it in a separate PR. I've included it in https://issues.redhat.com/browse/CRT-729 so we don't forget about it.